sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala

```diff
@@ -189,133 +265,83 @@ object ScalaJSPluginInternal {
     )
   }
 
+  /**
+   * Collect certain file types from a classpath.
+   *
+   * @param cp Classpath to collect from
+   * @param filter Filter for (real) files of interest (not in jars)
+   * @param collectJar Collect elements from a jar (called for all jars)
+   * @param collectFile Collect a single file. Params are the file and the
+   *     relative path of the file (to its classpath entry root).
+   * @return Collected elements attributed with physical files it originated
+   *     from (key: sourceFiles).
+   */
+  private def collectFromClasspath[T](cp: Def.Classpath, filter: FileFilter,
+      collectJar: VirtualJarFile => Seq[T],
+      collectFile: (File, String) => T): Attributed[Seq[T]] = {
+
+    val realFiles = Seq.newBuilder[File]
+    val results = Seq.newBuilder[T]
+
+    for (cpEntry <- Attributed.data(cp)) {
+      if (cpEntry.isFile && cpEntry.getName.endsWith(".jar")) {
+        realFiles += cpEntry
+        val vf = new FileVirtualBinaryFile(cpEntry) with VirtualJarFile
+        results ++= collectJar(vf)
+      } else if (cpEntry.isDirectory) {
+        for {
+          (file, relPath) <- Path.selectSubpaths(cpEntry, filter)
+        } {
+          realFiles += file
+          results += collectFile(file, relPath)
+        }
+      } else {
+        sys.error("Illegal classpath entry: " + cpEntry.getPath)
+      }
+    }
+
+    Attributed.blank(results.result()).put(sourceFiles, realFiles.result())
+  }
+
   val scalaJSConfigSettings: Seq[Setting[_]] = Seq(
       incOptions ~= scalaJSPatchIncOptions
   ) ++ (
-      scalajspSettings
+      scalajspSettings ++
+      stageKeys.flatMap((scalaJSStageSettings _).tupled) ++
+      dispatchKeySettings(scalaJSLinkingUnit) ++
+      dispatchKeySettings(scalaJSLinkedFile)
   ) ++ Seq(
+      /* Note: This cache only gets freed by the GC. Otherwise we'd need to
+       * intercept reloads in sbt.
```

Comment:
Ah I see. Maybe we should try to improve this in the future. sbt has an `onUnload` magical setting that can be used for that kind of stuff. Relying on finalizers is not really nice. Maybe file an issue for future improvement and reference it from here?